### PR TITLE
fix(codegen): make tile.slice a pure pto.subview view (fixes #1622)

### LINKED ISSUE #1622: [Bug] tensor.gather on local TileType input can overwrite source row during lowering
### Component

IR / `ConvertTensorToTileOps` / `tensor.gather` lowering for `TileType` inputs

### Description

`tensor.gather` can mutate or alias its local TileType input when the input is a local tensor lowered to a tile.

The repro below creates a local `[16, 256]` FP32 tensor `local_scores`, reads it with a scalar loop, then calls:

```python
gathered_tile = pl.gather(local_scores, dim=-1, index=idx)
```

The gather output itself is correct. However, reading the same `local_scores` again with the same scalar loop after `pl.gather` returns wrong values: row 0 behaves as if it has been overwritten by row 15.

This violates the expected read-only semantics of `gather`: neither the source input nor the index input should be modified by the operation.

This was observed on PyPTO main after #1609, where `tensor.gather` now accepts `TileType` input/index and no longer fails at compile time.

### Minimal Reproducer

Save this as `repro_tensor_gather_local_alias.py` and run it with the normal PyPTO runtime environment configured.

```python
import argparse

import pypto.language as pl
import torch
from pypto.runtime.runner import RunConfig


ROWS = 16
COLS = 256
TOPK = 8


@pl.jit
def repro_kernel(
    scores: pl.Tensor[[ROWS, COLS], pl.FP32],
    idx: pl.Tensor[[ROWS, TOPK], pl.INT32],
    before: pl.Out[pl.Tensor[[ROWS, TOPK], pl.FP32]],
    gathered: pl.Out[pl.Tensor[[ROWS, TOPK], pl.FP32]],
    after: pl.Out[pl.Tensor[[ROWS, TOPK], pl.FP32]],
):
    for _ in pl.spmd(1, name_hint="gather_alias_repro"):
        local_scores = pl.create_tensor([ROWS, COLS], dtype=pl.FP32)
        local_scores[:, :] = scores[:, :]

        local_idx = pl.create_tensor([ROWS, TOPK], dtype=pl.INT32)
        local_idx[:, :] = idx[:, :]

        before_tile = pl.full([ROWS, TOPK], dtype=pl.FP32, value=0.0)
        for r in pl.range(ROWS):
            for k in pl.range(TOPK):
                col_i32 = pl.read(local_idx, [r, k])
                col = pl.cast(col_i32, pl.INDEX)
                val = pl.read(local_scores, [r, col])
                pl.write(before_tile, [r, k], val)
        before[:, :] = before_tile

        gathered_tile = pl.gather(local_scores, dim=-1, index=idx)
        gathered[:, :] = gathered_tile

        after_tile = pl.full([ROWS, TOPK], dtype=pl.FP32, value=0.0)
        for r in pl.range(ROWS):
            for k in pl.range(TOPK):
                col_i32 = pl.read(local_idx, [r, k])
                col = pl.cast(col_i32, pl.INDEX)
                val = pl.read(local_scores, [r, col])
                pl.write(after_tile, [r, k], val)
        after[:, :] = after_tile


def make_scores() -> torch.Tensor:
    row = torch.arange(ROWS, dtype=torch.float32).reshape(ROWS, 1) * 1000.0
    col = torch.arange(COLS, dtype=torch.float32).reshape(1, COLS)
    return row + col


def make_idx() -> torch.Tensor:
    base = torch.tensor([7, 9, 3, 128, 255, 0, 64, 11], dtype=torch.int32)
    return torch.stack([(base + r * 13) % COLS for r in range(ROWS)])


def first_mismatch(actual: torch.Tensor, expected: torch.Tensor) -> str:
    neq = actual != expected
    if not bool(neq.any()):
        return ""
    flat_idx = int(torch.nonzero(neq.reshape(-1), as_tuple=False)[0].item())
    idx = torch.unravel_index(torch.tensor(flat_idx), actual.shape)
    idx_tuple = tuple(int(x) for x in idx)
    return f"{idx_tuple}: actual={actual[idx_tuple].item()} expected={expected[idx_tuple].item()}"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", default="a2a3sim", choices=["a2a3", "a2a3sim", "a5", "a5sim"])
    parser.add_argument("-d", "--device", type=int, default=0)
    args = parser.parse_args()

    scores = make_scores()
    idx = make_idx()
    expected = torch.gather(scores, dim=-1, index=idx.long())
    before = torch.zeros_like(expected)
    gathered = torch.zeros_like(expected)
    after = torch.zeros_like(expected)

    repro_kernel(
        scores,
        idx,
        before,
        gathered,
        after,
        config=RunConfig(platform=args.platform, device_id=args.device, dump_passes=True),
    )

    for name, actual in {"before": before, "gathered": gathered, "after": after}.items():
        ok = torch.equal(actual, expected)
        print(f"{name}: {'PASS' if ok else 'FAIL'}")
        if not ok:
            print(f"  first mismatch: {first_mismatch(actual, expected)}")


if __name__ == "__main__":
    main()
```

Run:

```bash
python repro_tensor_gather_local_alias.py -p a2a3sim
```

### Expected Behavior

All three outputs should match `torch.gather(scores, dim=-1, index=idx)`:

```text
before: PASS
gathered: PASS
after: PASS
```

`before` and `after` use the same scalar-read loop:

```python
val = pl.read(local_scores, [r, col])
pl.write(out_tile, [r, k], val)
```

The only operation between them is `pl.gather(local_scores, dim=-1, index=idx)`, which should not mutate `local_scores`.

### Actual Behavior

Observed result:

```text
before: PASS
gathered: PASS
after: FAIL
  first mismatch: (0, 0): actual=15007.0 expected=7.0
```

The input is constructed as:

```python
scores[row, col] = row * 1000 + col
```

Therefore `15007.0` means the `after` scalar-read path read row 15, col 7 where row 0, col 7 was expected. The scalar-read loop itself is not the cause, because the same loop passes before `pl.gather`.

### Preliminary Analysis

The generated PTO shows that the TileType-input `tensor.gather` path materializes each source row into a scratch tile before calling `pto.tgather`. The scratch tile for the sliced `[1, 256]` row aliases the same base address as the full `local_scores` tile:

```mlir
%local_scores__tile = pto.alloc_tile addr = %c96_i64 ... rows=16, cols=256
...
%0 = pto.alloc_tile addr = %c96_i64 ... rows=16, cols=256
...
%gather_inp_row = pto.alloc_tile addr = %c96_i64 ... rows=1, cols=256
%slice_view = pto.subview %0[%gather_lv, %c0_index] sizes [1, 256] ...
pto.tmov ins(%slice_view) outs(%gather_inp_row)
pto.tgather ins(%gather_inp_row, %gather_idx_row, %gather_row_tmp) outs(%gather_row)
```

Since `%gather_inp_row` starts at the same address as `local_scores`, every row materialization writes the current source row into `local_scores` row 0. The gather result is still correct in the same iteration because `pto.tgather` immediately consumes that scratch row. After the loop finishes, row 0 has been overwritten by the last materialized row, row 15. This explains the `after` mismatch:

```text
after[0, 0] reads local_scores[0, 7]
actual = 15007.0  # row 15, col 7
expected = 7.0    # row 0, col 7
```

This suggests that the TileType-input `tensor.gather` lowering needs to ensure its row materialization scratch does not alias the source input tile, or otherwise avoid materializing by writing into the source tile's address range.

### Environment

Observed with:

```text
PyPTO: b5341e805fae2c4ee0b3ec1615b59755fa1015f9
ptoas: 0.42
platform: a2a3sim
```

### Related

- #1609 fixed the compile-time `TileType` acceptance path for `tensor.gather`.
- This issue tracks a separate runtime correctness problem after the code can compile and execute.

### PR DESCRIPTION
**Pull request by YunjiQin**
## Summary

`tile.slice` codegen is now always a single `pto.subview`. When the 4-arg form carries an explicit `valid_shape`, the value is encoded directly into `pto.subview`'s `valid [%vr, %vc]` clause — the same syntax `tile.assemble` already uses. The previous materializing path (`pto.subview` + `pto.tmov ins(view) outs(result)` + `pto.set_validshape`) is removed entirely.

Removing this **eager** materializing path fixes issue #1622 at the root: `tile.slice` is declared `set_output_memory_inherit_input`, so the slice result's MemRef shares the source base. For dynamic offsets the alloc pass falls back to bare `src_base` (see `allocate_memory_addr_pass.cpp:346-353`'s "safe because `tile.slice` re-derives offset from operands" note — which held only for the pure-subview path, not for `tmov+set_validshape`). The previous per-row `tmov` therefore wrote into the source allocation's row 0, silently corrupting source rows after the gather loop completed. The slice op's own emission can no longer do this — it emits only a `pto.subview`, with no destination buffer to overlap the source.

**Scope note:** this removes the *eager* `tmov` path only. A separate, pre-existing lazy path remains — when a slice feeds `pto.tcolexpandmul`/`pto.tcolexpandadd`, those consumers materialize the subview operand via `pto.textract` into the slice's (inherited) result buffer, which can still alias the source for a dynamic-offset slice of a local tile. That path predates this change (#1217) and is unchanged here; it is tracked separately in #1640. No current model exercises it (real `col_expand` call sites slice GM tensor params → `tile.load`, not `tile.slice`).

The 3-arg path is unchanged in shape (still a single `pto.subview`); the lambda/early-return structure is collapsed and the now-unused `tmov`/`set_validshape` emission code is gone. `SubviewMaterializationInfo` is registered for all slices (3-arg and 4-arg) so `pto.tcolexpandmul`'s eager-materialization fallback continues to work (its pre-existing aliasing hazard is tracked in #1640).

A follow-up commit also fixes a per-dim `valid_shape` typing bug surfaced by the pypto-lib qwen3-14b `lm_head_store` kernel: for `valid_shape=[dynamic, static]`, the explicit-valid branch must reflect each dim's static/dynamic-ness independently (not promote both to dynamic), otherwise PTOAS rejects the static `valid_col` operand against a `v_col=?` result type.

## Changes

- `src/backend/common/pto_ops_common.cpp` — collapse `tile.slice` codegen to a single `pto.subview` branch; explicit `valid_shape` overrides `InferSubviewTileTypeComponents`'s `v_row`/`v_col` per-dim (no promote-both-to-dynamic) and feeds the subview's `valid [...]` clause.
- `tests/ut/codegen/test_pto_codegen.py` — invert two tests that previously mandated `tmov + set_validshape` (one renamed from `..._uses_set_validshape` to `..._uses_subview_valid`).
- `tests/ut/codegen/test_pto_codegen_ops.py` — invert `test_tile_slice_codegen_with_valid_shape` to expect the subview-`valid` form; add `test_tile_slice_mixed_dynamic_static_valid_shape_per_dim_type` (the `lm_head_store` per-dim regression).
- `tests/st/runtime/ops/test_gather.py` — add `test_gather_tile_input_source_unchanged`: builds a local tile via `pl.create_tensor` + `pl.assemble`, runs `pl.tensor.gather`, then echoes the local tile back out and asserts it is bit-identical to the input.

## Test plan

- [x] `tests/ut/codegen/` — 385 tests pass.
- [x] `tests/ut/ir/transforms/` — 1534 tests pass.
- [x] `tests/ut/{ir/operators,ir/printing,ir/parser,ir/statements,ir/expressions,ir/core,core,pass,language}/` — 2248 tests pass.
- [x] `tests/st/runtime/ops/test_gather.py::TestGatherIndex` on `a2a3sim` — 7/7 pass, including the new `test_gather_tile_input_source_unchanged` #1622 regression.
- [x] `examples.kernels.dyn_valid_shape` compiles for both partial-block (`vlen < BLOCK_COL`, dynamic valid) and full-block (`vlen == BLOCK_COL`) variants — i.e. the dynamic-`valid_shape` codegen path still works end-to-end.
- [x] Original #1622 reproducer (from issue body) on `a2a3sim`: `before: PASS / gathered: PASS / after: PASS`.
- [x] Negative-control: temporarily reverting the codegen change makes the new st regression fail with exactly 64/1024 mismatching elements (row 0 of the `[16, 64]` source) — matching the corruption pattern described in #1622.
- [x] pypto-lib qwen3-14b `decode_layer.py` (`lm_head_store`) compiles end-to-end (compile-only) with the rebuilt pypto + ptoas 0.43 — previously failed PTOAS with `'pto.subview' op expects result valid_shape[1] to match inferred/explicit valid_col`. Verified the old/new subview forms directly against ptoas with hand-written `.pto`.

Fixes #1622

Follow-up: #1640 (pre-existing lazy `pto.textract` materialization aliasing hazard, out of scope here).

### CODE REVIEW

File: src/backend/common/pto_ops_common.cpp [hunk 1]
```diff
@@ -3483,92 +3483,70 @@ void RegisterPTOOps(Backend& backend, const std::unordered_set<std::string>& exc
...
+    // reads physical tile dims from the operand type. MaterializeSubviewOperandIfNeeded
+    // will emit pto.textract on demand into result_target if such a consumer appears.
+    codegen::PTOCodegen::SubviewMaterializationInfo mat_info;
+    mat_info.source_ssa = src;
+    mat_info.source_type = src_type;
+    mat_info.row_off_ssa = row_off;
+    mat_info.col_off_ssa = col_off;
+    mat_info.materialize_target_ssa = result_target;
+    mat_info.materialize_target_type = result_type;
+    codegen.RegisterSubviewMaterialization(view_ssa, mat_info);
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/backend/common/pto_ops_common.cpp:3546]: Good catch — the hazard is real, and I confirmed it by construction: a dynamic-offset `tile.slice` of a **local tile** fed to `col_expand_mul` generates `pto.textract ins(%local, %r, 0) outs(%row)` where `%row`'s `alloc_tile addr` collapses onto `%local`'s base (inherited memory + dynamic-offset fallback), overwriting source row 0 — i.e. the same class as #1622, via the lazy path.

Two reasons I'm not fixing it in this PR:

1. **Pre-existing, not introduced here.** `materialize_target_ssa = result_target` and the `pto.textract` materialization are from #1217. This PR removes the *eager* `pto.tmov` path only; the 3-arg subview path (which registers the same `mat_info`) is byte-identical to before, and the 4-arg path I added doesn't feed `col_expand` anywhere.

2. **Currently unreachable + proper fix is a separate mechanism.** No model chains "dynamic-offset slice of a *local* tile → `col_expand` → re-read source". Real `col_expand` call sites slice GM tensor params, which lower to `tile.load` (fresh Vec load), not `tile.slice`. Fixing it correctly means materializing `pto.textract` into a non-inherited scratch buffer, which needs a codegen-layer non-aliased allocator (absent post-`AllocateMemoryAddr`) or an IR-level change to the slice's memory inheritance — out of scope for this PR.

Tracked as #1640 (with the constructed repro + generated PTO). The PR description and the refactor commit message now scope the "no source corruption" claim to the eager path explicitly. Resolving here.

**Review by lyfne123 — APPROVED** [review]
